In [1]:
import pandas as pd
import numpy as np
import warnings

from sklearn.metrics import precision_score
from sklearn.preprocessing import StandardScaler
from sklearn.neural_network import MLPClassifier

from sys import platform

if platform == "win32":
    path = 'C:/Users/olive/GitHub/f1-analytics/'
elif platform == "darwin":
    path = '~/Documents/GitHub/f1-analytics/'

warnings.filterwarnings("ignore", category=RuntimeWarning) 
pd.options.mode.chained_assignment = None  # default='warn'

%matplotlib inline

In [2]:
data = pd.read_csv(path+'data/ml_input.csv')
merged = pd.read_csv(path+'data/merged.csv')

In [3]:
### Season to test results

N = 2021

In [4]:
params_to_drop = ['season', 'round', 'podium', 'driver_points_from']

In [5]:
data.columns

Index(['season', 'round', 'podium', 'q_delta', 'starting_grid',
       'driver_points_before', 'constructor_points_before',
       'driver_points_from', 'driver_points_per', 'points_percentage',
       'driver_last_3', 'constructor_last_3', 'stage_q1', 'stage_q2',
       'stage_q3'],
      dtype='object')

In [6]:
df = data.copy()
df.podium = df.podium.map(lambda x: 1 if x == 1 else 0)

train = df[df.season < N]
X_train = train.drop(params_to_drop, axis=1)
y_train = train.podium

scaler = StandardScaler()
X_train = pd.DataFrame(scaler.fit_transform(X_train), columns = X_train.columns)

In [7]:
X_train

q_delta  starting_grid  driver_points_before  \
0    -0.960826      -1.294941             -0.679459   
1    -0.960826      -1.123101             -0.679459   
2    -0.872256      -0.092062             -0.679459   
3    -0.960826      -0.951261             -0.679459   
4     0.722864       0.767138             -0.679459   
...        ...            ...                   ...   
2737  0.329888       0.595298             -0.620844   
2738  0.413729       1.282658             -0.679459   
2739  0.164357       1.626338             -0.664805   
2740  0.297642       1.110818             -0.679459   
2741 -0.622024       1.454498              1.152244   

      constructor_points_before  driver_points_per  points_percentage  \
0                     -0.692428          -0.809546          -1.353452   
1                     -0.692428          -0.809546          -1.353452   
2                     -0.692428          -0.809546          -1.353452   
3                     -0.692428          -0.809546          -1.353452   
4                     -0.692428          -0.809546          -1.353452   
...                         ...                ...                ...   
2737                  -0.632629          -0.762335           0.225815   
2738                  -0.692428          -0.809546          -1.353452   
2739                  -0.670003          -0.797743          -0.300712   
2740                  -0.670003          -0.809546          -1.353452   
2741                   0.757701           0.829061           0.681592   

      driver_last_3  constructor_last_3  stage_q1  stage_q2  stage_q3  
0         -0.760054           -0.752773 -0.589424  -0.59167  1.034890  
1         -0.760054           -0.752773 -0.589424  -0.59167  1.034890  
2         -0.760054           -0.752773 -0.589424   1.69013 -0.966286  
3         -0.760054           -0.752773 -0.589424  -0.59167  1.034890  
4         -0.760054           -0.752773 -0.589424  -0.59167  1.034890  
...             ...                 ...       ...       ...       ...  
2737      -0.760054           -0.752773 -0.589424   1.69013 -0.966286  
2738      -0.760054           -0.752773  1.696573  -0.59167 -0.966286  
2739      -0.760054           -0.752773  1.696573  -0.59167 -0.966286  
2740      -0.760054           -0.670118  1.696573  -0.59167 -0.966286  
2741       1.649792            0.900340  1.696573  -0.59167 -0.966286  

[2742 rows x 11 columns]

In [8]:
def score_classification(model):
    score = 0
    for circuit in df[df.season == N]['round'].unique():

        test = df[(df.season == N) & (df['round'] == circuit)]
        X_test = test.drop(params_to_drop, axis=1)
        y_test = test.podium
        
        #scaling
        X_test = pd.DataFrame(scaler.transform(X_test), columns = X_test.columns)

        # make predictions
        prediction_df = pd.DataFrame(model.predict_proba(X_test), columns = ['proba_0', 'proba_1'])
        prediction_df['actual'] = y_test.reset_index(drop=True)
        prediction_df['predicted'] = prediction_df.proba_1.map(lambda x: 1 if x == prediction_df.proba_1.max() else 0)

        score += precision_score(prediction_df.actual, prediction_df.predicted)

    model_score = score / df[df.season == N]['round'].unique().max()
    return model_score

In [9]:
comparison_dict ={'model':[],
                  'hidden_layer_sizes': [],
                  'activation': [],
                  'solver': [],
                  'alpha': [],
                  'score': []
                  }

In [10]:
# Neural network

params={'hidden_layer_sizes': [(80,20,40,5), (75,25,50,10)], 
        'activation': ['identity', 'logistic', 'tanh', 'relu'], 
        'solver': ['lbfgs', 'sgd', 'adam'], 
        'alpha': np.logspace(-4,2,20)} 


for hidden_layer_sizes in params['hidden_layer_sizes']:
    for activation in params['activation']:
        for solver in params['solver']:
            for alpha in params['alpha']:
                model_params = (hidden_layer_sizes, activation, solver, alpha)
                model = MLPClassifier(
                    hidden_layer_sizes=hidden_layer_sizes,
                    activation=activation, 
                    solver=solver, 
                    alpha=alpha, 
                    random_state=1)

                model.fit(X_train, y_train)

                model_score = score_classification(model)

                comparison_dict['model'].append('neural_network_classifier')
                comparison_dict['hidden_layer_sizes'].append(hidden_layer_sizes)
                comparison_dict['activation'].append(activation)
                comparison_dict['solver'].append(solver)
                comparison_dict['alpha'].append(alpha)
                comparison_dict['score'].append(model_score)

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iteration

In [11]:
comparison_df = pd.DataFrame(comparison_dict).sort_values('score', ascending=False)

comparison_df.head()

model hidden_layer_sizes activation solver     alpha  \
210  neural_network_classifier    (80, 20, 40, 5)       relu    sgd  0.143845   
200  neural_network_classifier    (80, 20, 40, 5)       relu    sgd  0.000100   
209  neural_network_classifier    (80, 20, 40, 5)       relu    sgd  0.069519   
208  neural_network_classifier    (80, 20, 40, 5)       relu    sgd  0.033598   
207  neural_network_classifier    (80, 20, 40, 5)       relu    sgd  0.016238   

        score  
210  0.590909  
200  0.590909  
209  0.590909  
208  0.590909  
207  0.590909

model	hidden_layer_sizes	activation	solver	alpha	score
421	neural_network_classifier	(75, 25, 50, 10)	relu	lbfgs	0.000207	0.5
160	neural_network_classifier	(80, 20, 40, 5)	tanh	adam	0.000100	0.5
162	neural_network_classifier	(80, 20, 40, 5)	tanh	adam	0.000428	0.5
163	neural_network_classifier	(80, 20, 40, 5)	tanh	adam	0.000886	0.5
164	neural_network_classifier	(80, 20, 40, 5)	tanh	adam	0.001833	0.5

In [12]:
chosen_inputs = comparison_df.query('score >= 0.5')

chosen_inputs.to_csv(path+'parameters/neutralnetworkclassifier.csv')